model/rag/chat.go

```diff
@@ -0,0 +1,200 @@
+	cloned := *c
+	cloned.Messages = make([]ChatMessage, len(c.Messages))
+	copy(cloned.Messages, c.Messages)
+	return &cloned
+}
+func (c *ChatCompletion) Included() []jsonapi.Object             { return nil }
+func (c *ChatCompletion) Relationships() jsonapi.RelationshipMap { return nil }
+func (c *ChatCompletion) Links() *jsonapi.LinksList              { return nil }
+
+var _ jsonapi.Object = (*ChatCompletion)(nil)
+
+type QueryMessage struct {
+	Task  string `json:"task"`
+	DocID string `json:"doc_id"`
+}
+
+func Chat(inst *instance.Instance, payload ChatPayload) (*ChatCompletion, error) {
+	var chat ChatCompletion
+	err := couchdb.GetDoc(inst, consts.ChatCompletions, payload.ChatCompletionID, &chat)
+	if couchdb.IsNotFoundError(err) {
+		chat.DocID = payload.ChatCompletionID
+	} else if err != nil {
+		return nil, err
+	}
+	msg := ChatMessage{Role: HumanRole, Content: payload.Query}
+	chat.Messages = append(chat.Messages, msg)
+	if chat.DocRev == "" {
+		err = couchdb.CreateNamedDocWithDB(inst, &chat)
+	} else {
+		err = couchdb.UpdateDoc(inst, &chat)
+	}
+	if err != nil {
+		return nil, err
+	}
+	query, err := job.NewMessage(&QueryMessage{
+		Task:  "chat-completion",
+		DocID: chat.DocID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	_, err = job.System().PushJob(inst, &job.JobRequest{
+		WorkerType: "rag-query",
+		Message:    query,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &chat, nil
+}
+
+func Query(inst *instance.Instance, logger logger.Logger, query QueryMessage) error {
+	var chat ChatCompletion
+	err := couchdb.GetDoc(inst, consts.ChatCompletions, query.DocID, &chat)
+	if err != nil {
+		return err
+	}
+	msg := chat.Messages[len(chat.Messages)-1]
+	payload := map[string]interface{}{
+		"q":      msg.Content,
+		"stream": true,
+	}
+
+	res, err := callRAGQuery(inst, payload)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
+		return fmt.Errorf("POST status code: %d", res.StatusCode)
+	}
+
+	var completion string
+	err = foreachSSE(res.Body, func(event map[string]interface{}) {
+		switch event["object"] {
+		case "delta", "done":
+			content, _ := event["content"].(string)
+			completion += content
+			delta := couchdb.JSONDoc{
+				Type: consts.ChatEvents,
+				M:    event,
+			}
+			go realtime.GetHub().Publish(inst, realtime.EventCreate, &delta, nil)
+		default:
+			// We can ignore done events
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	answer := ChatMessage{
+		Role:    AIRole,
+		Content: completion,
+	}
+	chat.Messages = append(chat.Messages, answer)
+	return couchdb.UpdateDoc(inst, &chat)
+}
+
+func callRAGQuery(inst *instance.Instance, payload map[string]interface{}) (*http.Response, error) {
+	ragServer := inst.RAGServer()
+	if ragServer.URL == "" {
+		return nil, errors.New("no RAG server configured")
+	}
+	u, err := url.Parse(ragServer.URL)
+	if err != nil {
+		return nil, err
+	}
+	u.Path = fmt.Sprintf("/query/%s", inst.Domain)
```

Comment:
I wonder if `/query` is not too generic, as we might have other endpoints in the future. Maybe `/chat` would be more precise?

Reply:
I was also hesitating between a generic endpoint that accepts a parameter in the JSON (like `task: chat`), or having an endpoint per task.